src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/MeanQualityByCycleSpark.java

```diff
@@ -0,0 +1,273 @@
+        HistogramGeneratorPair addRead(GATKRead read){
+            q.addRecord(read);
+            oq.addRecord(read);
+            return this;
+        }
+        HistogramGeneratorPair merge(HistogramGeneratorPair other){
+            q.merge(other.q);
+            oq.merge(other.oq);
+            return this;
+        }
+
+    }
+
+    @Override
+    protected void runPipeline(final JavaSparkContext ctx) {
+        if ( readArguments.getReadFilesNames().size() != 1 ) {
+            throw new UserException("This tool only accepts a single bam/sam/cram as input");
+        }
+        final String bam = readArguments.getReadFilesNames().get(0);
+        final ReadsSparkSource readSource = new ReadsSparkSource(ctx);
+        final SAMFileHeader readsHeader = ReadsSparkSource.getHeader(ctx, bam);
+        setup(readsHeader);
+
+        /**
+         * If no intervals specified - use all reads, mapped and unmapped.
+         */
+        final List<SimpleInterval> intervals = intervalArgumentCollection.intervalsSpecified() ? intervalArgumentCollection.getIntervals(readsHeader.getSequenceDictionary())
+                : null;
+
+        final JavaRDD<GATKRead> reads = readSource.getParallelReads(bam, intervals);
+        final JavaRDD<GATKRead> filteredReads = reads.filter(makeReadFilter(PF_READS_ONLY, ALIGNED_READS_ONLY));
+        final HistogramGeneratorPair aggregate = filteredReads.aggregate(new HistogramGeneratorPair(),
+                (hgp, read) -> hgp.addRead(read),
+                (hgp1, hgp2) -> hgp1.merge(hgp2));
+        finish(aggregate.q, aggregate.oq, bam);
+    }
+
+    private static Function<GATKRead, Boolean> makeReadFilter(final boolean pf_read_only, final boolean aligned_reads_only) {
+        return rec -> (!pf_read_only || !rec.failsVendorQualityCheck()) &&
+                      (!aligned_reads_only || !rec.isUnmapped()) &&
+                      !rec.isSecondaryAlignment() &&
+                      !rec.isSupplementaryAlignment();
+    }
+
+    private void setup(final SAMFileHeader header) {
+        if (PRODUCE_PLOT) {
+            IOUtil.assertFileIsWritable(CHART_OUTPUT);
+        }
+        // If we're working with a single library, assign that library's name
+        // as a suffix to the plot title
+        final List<SAMReadGroupRecord> readGroups = header.getReadGroups();
+        if (readGroups.size() == 1) {
+            plotSubtitle = StringUtil.asEmptyIfNull(readGroups.get(0).getLibrary());
+        }
+    }
+
+    private void finish(final HistogramGenerator q, final HistogramGenerator oq, final String inputFileName) {
+        // Generate a "Histogram" of mean quality and write it to the file
+        final MetricsFile<?,Integer> metrics = getMetricsFile();
+        metrics.addHistogram(q.getMeanQualityHistogram());
+        if (!oq.isEmpty()) metrics.addHistogram(oq.getMeanQualityHistogram());
+        final File outFile = new File(out);
+        metrics.write(outFile);
```

Comment:
This won't write to GCS or HDFS.  Can you rework to use `BucketUtils.createFile()` to get an output stream and then use the `metrics.write` overload that takes `Writer` instead?